File: showcase/integrations/langgraph-typescript/src/agent/byoc-hashbrown.ts
/**
 * LangGraph TypeScript agent backing the byoc-hashbrown demo.
 *
 * Port of `langgraph-python/src/agents/byoc_hashbrown_agent.py`.
 *
 * Emits hashbrown-shaped structured output (`<ui>...</ui>`) that the frontend
 * renderer (`src/app/demos/byoc-hashbrown/hashbrown-renderer.tsx`) parses
 * progressively via `@hashbrownai/react`.
 *
 * A minimal single-node StateGraph (no tools) — the system prompt teaches
 * the small component catalog exposed by the frontend kit.
 */

import { RunnableConfig } from "@langchain/core/runnables";
import { SystemMessage } from "@langchain/core/messages";
import {
  MemorySaver,
  START,
  END,
  StateGraph,
  Annotation,
} from "@langchain/langgraph";
import { ChatOpenAI } from "@langchain/openai";
import { CopilotKitStateAnnotation } from "@copilotkit/sdk-js/langgraph";

// `@hashbrownai/react`'s `useJsonParser(content, kit.schema)` expects the
// agent to stream a JSON object literal matching `kit.schema` — NOT the
// `<ui>...</ui>` XML-style examples shown inside `useUiKit({ examples })`.
// Those XML examples are the hashbrown prompt DSL that hashbrown compiles
// into a schema description when driving the LLM directly. Since this demo
// drives the LLM via langgraph, we mirror hashbrown's wire format — a
// `{ ui: [...] }` envelope — instead.
//
// Every node is a single-key object `{ tagName: { props: {...} } }`. Tag
// names and prop schemas match `useSalesDashboardKit()` in
// `hashbrown-renderer.tsx`. `pieChart`/`barChart` receive `data` as a
// JSON-encoded string to keep the schema stable under partial streaming.
const BYOC_HASHBROWN_SYSTEM_PROMPT = `You are a sales analytics assistant that replies by emitting a single JSON
object consumed by a streaming JSON parser on the frontend.

ALWAYS respond with a single JSON object of the form:

{
  "ui": [
    { <componentName>: { "props": { ... } } },
    ...
  ]
}

Do NOT wrap the response in code fences. Do NOT include any preface or
explanation outside the JSON object. The response MUST be valid JSON.

Available components and their prop schemas:

- "metric": { "props": { "label": string, "value": string } }
    A KPI card. \`value\` is a pre-formatted string like "$1.2M" or "248".

- "pieChart": { "props": { "title": string, "data": string } }
    A donut chart. \`data\` is a JSON-encoded STRING (embedded JSON) of an
    array of {label, value} objects with at least 3 segments, e.g.
    "data": "[{\\"label\\":\\"Enterprise\\",\\"value\\":600000}]".

- "barChart": { "props": { "title": string, "data": string } }
    A vertical bar chart. \`data\` is a JSON-encoded STRING of an array of
    {label, value} objects with at least 3 bars, typically time-ordered.

- "dealCard": { "props": { "title": string, "stage": string, "value": number } }
    A single sales deal. \`stage\` MUST be one of: "prospect", "qualified",
    "proposal", "negotiation", "closed-won", "closed-lost". \`value\` is a
    raw number (no currency symbol or comma).

- "Markdown": { "props": { "children": string } }
    Short explanatory text. Use for section headings and brief summaries.
    Standard markdown is supported in \`children\`.

Rules:
- Always produce plausible sample data when the user asks for a dashboard or
  chart — do not refuse for lack of data.
- Prefer 3-6 rows of data in charts; keep labels short.
- Use "Markdown" for short headings or linking sentences between visual
  components. Do not emit long prose.
- Do not emit components that are not listed above.
- \`data\` props on charts MUST be a JSON STRING — escape inner quotes.

Example response (sales dashboard):
{"ui":[{"Markdown":{"props":{"children":"## Q4 Sales Summary"}}},{"metric":{"props":{"label":"Total Revenue","value":"$1.2M"}}},{"metric":{"props":{"label":"New Customers","value":"248"}}},{"pieChart":{"props":{"title":"Revenue by Segment","data":"[{\\"label\\":\\"Enterprise\\",\\"value\\":600000},{\\"label\\":\\"SMB\\",\\"value\\":400000},{\\"label\\":\\"Startup\\",\\"value\\":200000}]"}}},{"barChart":{"props":{"title":"Monthly Revenue","data":"[{\\"label\\":\\"Oct\\",\\"value\\":350000},{\\"label\\":\\"Nov\\",\\"value\\":400000},{\\"label\\":\\"Dec\\",\\"value\\":450000}]"}}}]}
`;

const AgentStateAnnotation = Annotation.Root({
  ...CopilotKitStateAnnotation.spec,
});

type AgentState = typeof AgentStateAnnotation.State;

async function chatNode(state: AgentState, config: RunnableConfig) {
  // Force JSON-object output mode. The frontend's `useJsonParser` bails
  // to `null` on any non-JSON prefix (code fences, prose preamble, etc.),
  // so locking the model to JSON at the API layer keeps the wire
  // contract honest. Passed via `modelKwargs` so it survives the
  // LangChain → OpenAI chat-completions mapping.
  const model = new ChatOpenAI({
    model: "gpt-4o-mini",
    modelKwargs: { response_format: { type: "json_object" } },
  });
  const response = await model.invoke(
    [
      new SystemMessage({ content: BYOC_HASHBROWN_SYSTEM_PROMPT }),
      ...state.messages,
    ],
    config,
  );
  return { messages: response };
}

const workflow = new StateGraph(AgentStateAnnotation)
  .addNode("chat_node", chatNode)
  .addEdge(START, "chat_node")
  .addEdge("chat_node", END);

const memory = new MemorySaver();

export const graph = workflow.compile({ checkpointer: memory });

File: showcase/integrations/strands/src/agents/byoc_hashbrown.py
"""Strands agent specialization for the byoc-hashbrown demo (Wave 2).

The Strands showcase (as documented in `PARITY_NOTES.md`) historically ships
a single shared Strands agent (`src/agents/agent.py`) registered under many
AG-UI agent names. The byoc-hashbrown demo additionally requires the LLM to
emit a strict **hashbrown JSON envelope** (NOT XML) that
`@hashbrownai/react`'s `useJsonParser` + `useUiKit` can progressively parse.

Wire format
-----------
The renderer (`src/app/demos/byoc-hashbrown/hashbrown-renderer.tsx`) consumes
a JSON object shaped like the hashbrown schema itself — not the XML example
DSL used inside `useUiKit({ examples })`. Because this demo drives the LLM
via a Strands Agent (not via hashbrown's own `useUiChat`), we must emit the
raw schema wire format:

    {
      "ui": [
        { "metric":   { "props": { "label": "...", "value": "..." } } },
        { "pieChart": { "props": { "title": "...", "data": "[{...}]" } } },
        { "barChart": { "props": { "title": "...", "data": "[{...}]" } } },
        { "dealCard": { "props": { "title": "...", "stage": "prospect", "value": 100000 } } },
        { "Markdown": { "props": { "children": "## heading\\nbody" } } }
      ]
    }

Every node is a single-key object `{tagName: {props: {...}}}`. The tag names
and prop schemas match `useSalesDashboardKit()` in `hashbrown-renderer.tsx`.
`pieChart` / `barChart` receive `data` as a JSON-encoded string (kept as a
string so the schema is stable under partial streaming).
"""

BYOC_HASHBROWN_SYSTEM_PROMPT = """\
You are a sales analytics assistant that replies by emitting a single JSON
object consumed by a streaming JSON parser on the frontend.

ALWAYS respond with a single JSON object of the form:

{
  "ui": [
    { <componentName>: { "props": { ... } } },
    ...
  ]
}

Do NOT wrap the response in code fences. Do NOT include any preface or
explanation outside the JSON object. The response MUST be valid JSON.

Available components and their prop schemas:

- "metric": { "props": { "label": string, "value": string } }
    A KPI card. `value` is a pre-formatted string like "$1.2M" or "248".

- "pieChart": { "props": { "title": string, "data": string } }
    A donut chart. `data` is a JSON-encoded STRING (embedded JSON) of an
    array of {label, value} objects with at least 3 segments.

- "barChart": { "props": { "title": string, "data": string } }
    A vertical bar chart. `data` is a JSON-encoded STRING of an array of
    {label, value} objects with at least 3 bars, typically time-ordered.

- "dealCard": { "props": { "title": string, "stage": string, "value": number } }
    A single sales deal. `stage` MUST be one of: "prospect", "qualified",
    "proposal", "negotiation", "closed-won", "closed-lost". `value` is a
    raw number (no currency symbol or comma).

- "Markdown": { "props": { "children": string } }
    Short explanatory text. Use for section headings and brief summaries.

Rules:
- Always produce plausible sample data when the user asks for a dashboard or
  chart -- do not refuse for lack of data.
- Prefer 3-6 rows of data in charts; keep labels short.
- Do not emit components that are not listed above.
- `data` props on charts MUST be a JSON STRING -- escape inner quotes.
"""


def build_byoc_hashbrown_agent():
    """Build a StrandsAgent configured with the byoc-hashbrown system prompt.

    Returns an ``ag_ui_strands.StrandsAgent`` wrapper (mirrors
    ``build_voice_agent``) so it can be mounted by ``create_strands_app`` and
    exposed as a dedicated AG-UI endpoint. agent_server.py mounts it at
    ``/byoc-hashbrown`` and the declarative-hashbrown route proxies there.
    The agent takes no tools; it is a pure structured-output generator.
    """
    # Deferred imports so this module remains importable even when the
    # agent_server import-order patches (see agent_server.py) haven't been
    # applied yet. Mirrors build_voice_agent: ``from strands import Agent``
    # plus the ag_ui_strands wrapper. The OpenAI model is built via the
    # shared agents.agent._build_model factory so we don't re-resolve the
    # ``strands.models.openai`` submodule independently.
    from strands import Agent
    from ag_ui_strands import StrandsAgent

    from agents.agent import _build_model

    strands_agent = Agent(
        model=_build_model(),
        system_prompt=BYOC_HASHBROWN_SYSTEM_PROMPT,
        tools=[],
    )
    return StrandsAgent(
        agent=strands_agent,
        name="byoc_hashbrown",
        description="Hashbrown UI-kit envelope generator for the declarative-hashbrown demo.",
    )
